campaignion_tracking/package/src/ga4.js

```diff
@@ -0,0 +1,426 @@
+    data = this.callChangeHook(eventName, data, this._context)
+
+    const changedNewProduct = data.addData.ecommerce.add.products[0]
+    const changedCurrentProduct = data.removeData.ecommerce.remove.products[0]
+
+    // Only change something or send an event if the donation products differ.
+    // Compare *after* any changes.
+    if (productIsEqual(changedCurrentProduct, changedNewProduct)) {
+      this.printDebug('(handle)', eventName, 'same product', eventData, context)
+      return
+    }
+
+    this._context.donation.product = changedNewProduct
+    this.saveToStorage('context', this._context)
+
+    if (data.pushRemove) {
+      this.dataLayer.push(data.removeData)
+    }
+    this.dataLayer.push(data.addData)
+  }
+
+  /**
+   * Handle "checkoutBegin".
+   *
+   * Event data: { product, currencyCode }
+   *
+   * @param {String} eventName the event name
+   * @param {object} eventData data of the event
+   * @param {object} context context of the event
+   */
+  handle_checkoutBegin (eventName, eventData, context) {
+    this.printDebug('(handle)', eventName, eventData, context)
+    this.updateContext(context)
+    const product = eventData.product || this._context.donation.product || {}
+    const currencyCode = eventData.currencyCode || this._context.donation.currencyCode || null
+    let data = {
+      event: 'checkoutBegin',
+      ecommerce: {
+        checkout: {
+          actionField: { step: 1 }, // begin == 1
+          products: [product]
+        }
+      }
+    }
+    if (currencyCode) {
+      data.ecommerce.currencyCode = currencyCode
+    }
+    // Allow others to modify the data being sent to Google Analytics.
+    data = this.callChangeHook(eventName, data, this._context)
+    this.dataLayer.push(data)
+  }
+
+  /**
+   * Handle "checkoutEnd".
+   *
+   * Event data: { product, currencyCode }
+   *
+   * @param {String} eventName the event name
+   * @param {object} eventData data of the event
+   * @param {object} context context of the event
+   */
+  handle_checkoutEnd (eventName, eventData, context) {
+    this.printDebug('(handle)', eventName, eventData, context)
+    this.updateContext(context)
+    const product = eventData.product || this._context.donation.product || {}
+    const currencyCode = eventData.currencyCode || this._context.donation.currencyCode || null
+    let data = {
+      event: 'checkoutEnd',
+      ecommerce: {
+        checkout: {
+          actionField: { step: 2 }, // end == 2
+          products: [product]
+        }
+      }
+    }
+    if (currencyCode) {
+      data.ecommerce.currencyCode = currencyCode
+    }
+    // Allow others to modify the data being sent to Google Analytics.
+    data = this.callChangeHook(eventName, data, this._context)
+    this.dataLayer.push(data)
+  }
+
+  /**
+   * Handle "donationSuccess".
+   *
+   * The event data needs to include an transaction id.
+   * Falls back to use a random number.
```

Comment:
This comment contradicts itself: Either it is required or there is a default. It can’t be both.

Reply:
I don’t think so. The transaction id is required by google, not by this method. If no id is passed, the method will create a random id to send in the event.

Reply:
I don’t think it’s obvious that this comment refers to GA’s needs instead of the function’s parameters.

Reply:
I didn't write the comment and I find it pretty clear. But you’re welcome to make a suggestion for improvement!